### PR TITLE
Chore/python3

### PR DESCRIPTION
**Pull request by paulineribeyre**
- updated dockerfile to use the python 3 base image
- did NOT update shibboleth dockerfile to work with python 3
- ran 2to3 for python 2 to 3 syntax updates
- encoding fixes: lots of encoding issues due to all python 3 strings now being unicode; may have missed some…
- Exception.message does not exist in python 3, so the custom Gen3 exceptions (CirrusNotFound, GoogleAPIError, etc) that inherit from Exception don't have a `message` attribute anymore
- removed `enum34` dependency because it is not compatible with the standard library's `enum` on python >3.4
- removed conditional imports for python 2/3 compatibility since Fence now only works on python 3
- fixed test logic `test_update_user_service_account_success`

TODO:
pin versions after PRs are merged:
- https://github.com/uc-cdis/cirrus/pull/70
- https://github.com/uc-cdis/storage-client/pull/23
- https://github.com/uc-cdis/userdatamodel/pull/60
- https://github.com/uc-cdis/cdis-python-utils/pull/39
- https://github.com/uc-cdis/authutils/pull/36
- https://github.com/uc-cdis/gen3-qa/pull/144

### Breaking Changes
- Python 3 instead of python 2

### Dependency updates
- Remove unused flask-postgres-session dependency
- Remove enum34 dependency
- Use python 3 versions of cirrus (1.0.0), storage-client, cdisutilstest (1.0.0), userdatamodel (2.0.0), cdispyutils (1.0.0) and authutils (4.0.0)

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
This PR contains code that is not formatted correctly according to [`black`](https://github.com/ambv/black). Run `black` on your code before merging.

<details>
<summary>Expand the full diff to see formatting changes</summary>

```diff
--- fence/__init__.py
+++ blackened
@@ -16,11 +16,11 @@
 from fence.oidc.server import server
 from fence.rbac.client import ArboristClient
 from fence.resources.aws.boto_manager import BotoManager
 from fence.resources.openid.google_oauth2 import GoogleOauth2Client as GoogleClient
 from fence.resources.openid.microsoft_oauth2 import (
-    MicrosoftOauth2Client as MicrosoftClient
+    MicrosoftOauth2Client as MicrosoftClient,
 )
 from fence.resources.openid.orcid_oauth2 import OrcidOauth2Client as ORCIDClient
 from fence.resources.storage import StorageManager
 from fence.resources.user.user_session import UserSessionInterface
 from fence.error_handler import get_error_response

--- fence/blueprints/google.py
+++ blackened
@@ -456,20 +456,16 @@
                 sa.google_project_id, sa.email, sa.project_access
             )
 
         except CirrusNotFound as exc:
             return (
-                "Can not update the service accout {}. Detail {}".format(
-                    sa.email, exc
-                ),
+                "Can not update the service accout {}. Detail {}".format(sa.email, exc),
                 404,
             )
         except GoogleAPIError as exc:
             return (
-                "Can not update the service accout {}. Detail {}".format(
-                    sa.email, exc
-                ),
+                "Can not update the service accout {}. Detail {}".format(sa.email, exc),
                 400,
             )
         except Exception:
             return (" Can not delete the service account {}".format(sa.email), 500)
 
@@ -500,20 +496,16 @@
                 service_account_email, google_project_id
             )
             force_delete_service_account(service_account_email)
         except CirrusNotFound as exc:
             return (
-                "Can not remove the service accout {}. Detail {}".format(
-                    id_, exc
-                ),
+                "Can not remove the service accout {}. Detail {}".format(id_, exc),
                 404,
             )
         except GoogleAPIError as exc:
             return (
-                "Can not remove the service accout {}. Detail {}".format(
-                    id_, exc
-                ),
+                "Can not remove the service accout {}. Detail {}".format(id_, exc),
                 400,
             )
         except Exception:
             return (" Can not delete the service account {}".format(id_), 500)
 

--- fence/oidc/endpoints.py
+++ blackened
@@ -54,11 +54,13 @@
 
         # The stored client secret is hashed, so hash the secret from basic
         # authorization header to check against stored hash.
         hashed = client.client_secret
         if (
-            bcrypt.hashpw(client_secret.encode("utf-8"), hashed.encode("utf-8")).decode("utf-8")
+            bcrypt.hashpw(client_secret.encode("utf-8"), hashed.encode("utf-8")).decode(
+                "utf-8"
+            )
             != hashed
         ):
             logger.debug("client secret hash does not match stored secret hash")
             raise InvalidClientError(uri=self.uri)
 

--- fence/oidc/grants/authorization_code_grant.py
+++ blackened
@@ -133,11 +133,13 @@
             client_id, client_secret = client_params
             client = self.get_and_validate_client(client_id)
             # Client secrets are stored as hash.
             hashed = client.client_secret
             if (
-                bcrypt.hashpw(client_secret.encode("utf-8"), hashed.encode("utf-8")).decode("utf-8")
+                bcrypt.hashpw(
+                    client_secret.encode("utf-8"), hashed.encode("utf-8")
+                ).decode("utf-8")
                 != hashed
             ):
                 raise InvalidClientError(uri=self.uri)
 
             return client

--- fence/resources/aws/boto_manager.py
+++ blackened
@@ -148,26 +148,22 @@
             for group in groups:
                 if group["GroupName"] in group_names:
                     res[group["GroupName"]] = group
         except Exception as ex:
             self.logger.exception(ex)
-            raise UserError(
-                "Fail to get list of groups {}: {}".format(group_names, ex)
-            )
+            raise UserError("Fail to get list of groups {}: {}".format(group_names, ex))
         return res
 
     def create_user_group(self, group_name, path=None):
         try:
             group = self.iam.create_group(GroupName=group_name)["Group"]
             self.__create_policy__(
                 group_name, self.__get_policy_document_by_group_name__(group_name)
             )
         except Exception as ex:
             self.logger.exception(ex)
-            raise UserError(
-                "Fail to create group {}: {}".format(group_name, ex)
-            )
+            raise UserError("Fail to create group {}: {}".format(group_name, ex))
         return group
 
     def __get_policy_document_by_group_name__(self, group_name):
         """
         Getting policy document from config file and replace with actual value (same as project name)

--- fence/resources/openid/google_oauth2.py
+++ blackened
@@ -28,11 +28,13 @@
         Get authorization uri from discovery doc
         """
         authorization_endpoint = self.get_value_from_discovery_doc(
             "authorization_endpoint", "https://accounts.google.com/o/oauth2/v2/auth"
         )
-        uri, _ = self.session.create_authorization_url(authorization_endpoint, prompt="login")
+        uri, _ = self.session.create_authorization_url(
+            authorization_endpoint, prompt="login"
+        )
 
         return uri
 
     def get_user_id(self, code):
         """

--- fence/resources/openid/microsoft_oauth2.py
+++ blackened
@@ -28,11 +28,13 @@
         """
         authorization_endpoint = self.get_value_from_discovery_doc(
             "authorization_endpoint",
             "https://login.microsoftonline.com/organizations/oauth2/v2.0/authorize",
         )
-        uri, _ = self.session.create_authorization_url(authorization_endpoint, prompt="login")
+        uri, _ = self.session.create_authorization_url(
+            authorization_endpoint, prompt="login"
+        )
 
         return uri
 
     def get_user_id(self, code):
         try:
```
</details></br>

This formatting comment was generated automatically by a script in [uc-cdis/wool](https://github.com/uc-cdis/wool).

File: fence/resources/google/validity.py [hunk 1]
```diff
@@ -88,9 +88,6 @@ def __len__(self):
     def __bool__(self):
         return self._valid
 
-    def __nonzero__(self):
```
Comment (Avantol13) [inline fence/resources/google/validity.py]: we need the Python 3 equivilent of this, which I believe is `__bool__`. these objects needs to be able to go into a `bool(obj)` call and get true or false

File: fence/resources/user/__init__.py [hunk 2]
```diff
@@ -124,7 +124,7 @@ def get_user_info(current_session, username):
 
 def _get_optional_userinfo(user, claims):
     info = {}
-    for claim, claim_request in claims.iteritems():
+    for claim in claims:
```
Comment (Avantol13) [inline fence/resources/user/__init__.py]: I think this is still a dictionary? it at least defaults to one just above this

File: fence/sync/utils.py [hunk 3]
```diff
@@ -138,7 +141,7 @@ def insert_segment(current, segment):
 
                 # In [3]: list(map(lambda x: x["name"] == "b", xs)).index(True)
                 # Out[3]: 1
-                i = list(map(lambda c: c["name"] == segment, current)).index(True)
+                i = list([c["name"] == segment for c in current]).index(True)
```
Comment (Avantol13) [inline fence/sync/utils.py]: can you update the above example? 

`In [2]:` should now be `list([c["name"] == "b" for c in xs])`
`In [3]:` should now be `list([c["name"] == "b" for c in xs]).index(True)`

File: tests/google/test_access_utils.py [hunk 4]
```diff
@@ -332,8 +326,8 @@ def test_update_user_service_account_success(cloud_manager, db_session, setup_da
     project = db_session.query(Project).filter_by(auth_id="test_auth_1").first()
 
     project_ids = [
-        project.id
-        for project in (
+        item.project_id
+        for item in (
```
Comment (Avantol13) [inline tests/google/test_access_utils.py]: :confused: how was this working before?

File: tests/test_import_fence.py [hunk 5]
```diff
@@ -22,15 +22,15 @@ def reload_modules(module_name):
...
     fence_submodules = [
-        module for module in sys.modules.keys() if module.startswith(module_name)
+        module for module in list(sys.modules.keys()) if module.startswith(module_name)
     ]
     for module in fence_submodules:
         if sys.modules[module]:
             # SQLAlchemy gets upset when a table is loaded twice, so ignore
             # that.
             try:
                 # NOTE: in python3 this should become ``importlib.reload``
```
Comment (Avantol13) [inline tests/test_import_fence.py]:
```suggestion
```

File: tests/test_logout.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import mock
-import urlparse
-import urllib
+import urllib.request, urllib.parse, urllib.error
+from fence.resources.storage.cdis_jwt import create_session_token
```
Comment (Avantol13) [inline tests/test_logout.py]: is this needed? it doesn't look like it was there before

File: fence/jwt/keys.py [hunk 7]
```diff
@@ -181,8 +181,13 @@ def public_key_to_jwk(self):
         Return:
             dict: JWK representation of the public key
         """
-        n, e = _rsa_public_numbers(self.public_key)
+        # n, e = _rsa_public_numbers(self.public_key)
         jwk_dict = jwk.construct(self.public_key, algorithm="RS256").to_dict()
+        for k in jwk_dict:  # bytes to string
+            try:
+                jwk_dict[k] = jwk_dict[k].decode("utf-8")
+            except AttributeError:
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply fence/jwt/keys.py]: some values in `jwk_dict` are bytes, others are strings. strings can't be decoded
`AttributeError: 'str' object has no attribute 'decode'`

File: fence/resources/google/validity.py (hunk 1 above)
Comment (paulineribeyre) [reply fence/resources/google/validity.py]: `__bool__` already exists right above this deleted function

File: fence/resources/user/__init__.py (hunk 2 above)
Comment (paulineribeyre) [reply fence/resources/user/__init__.py]: yes, but codacy complained about `claim_request` not being used so i just cleaned it up by iterating over the keys

File: tests/test_logout.py (hunk 6 above)
Comment (paulineribeyre) [reply tests/test_logout.py]: it's duplicated :o looks like the imports have been messed up during a rebase to master, thanks will fix

File: tests/google/test_access_utils.py (hunk 4 above)
Comment (paulineribeyre) [reply tests/google/test_access_utils.py]:
```
project = query_on_Project.first()
project_ids = [ project.id for project in (query_on_ServiceAccountAccessPrivilege) ]
assert project_ids[0] == project.id
```
in python 2, `project` outside of the loop gets overwritten by `project` inside of the loop. since `len(project_ids)==1`, `project_ids[0]` == id of the last overwritten `project` value. so the assertion is true (it compares a "ServiceAccountAccessPrivilege id" to itself)

in python3, `project` is not overwritten (https://www.python.org/dev/peps/pep-0289/) and the assertion compares a "Project id" to a "ServiceAccountAccessPrivilege id" -> fails

tldr: it gave me a headache

File: fence/jwt/keys.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Avantol13) [reply fence/jwt/keys.py]: yeah, that might be more readable. or if you just add a comment clarifying this I think it's okay the way it is

File: fence/resources/google/validity.py (hunk 1 above)
Comment (Avantol13) [reply fence/resources/google/validity.py]: :man_facepalming:

File: tests/google/test_access_utils.py (hunk 4 above)
Comment (Avantol13) [reply tests/google/test_access_utils.py]: woow... that's a really strange one. good catch

File: README.md [hunk 8]
```diff
@@ -101,6 +101,7 @@ At the moment, supported IDPs include:
   - InCommon
   - eduGAIN
 
+Note: the Shibboleth dockerfile image is at https://quay.io/repository/cdis/fence-shib and is NOT compatible with python 3/the latest fence.
```
Comment (Avantol13) [inline README.md]: is there a specific reason for this?

Comment (paulineribeyre) [reply README.md]: i didn't update the shibboleth dockerfile to work with python 3 code for now, I discussed with Rudy and he said master fence probably won't be deployed to the commons that's using this dockerfile
